go-controller/pkg/ovn/zone_interconnect/zone_ic_handler.go

```diff
@@ -791,6 +794,69 @@ func (zic *ZoneInterconnectHandler) getStaticRoutes(ipPrefixes []*net.IPNet, nex
 	return staticRoutes
 }
 
+// deleteStaleStaticRoutes removes static routes referencing old join or transit switch
+// IPs as either IPPrefix or NextHop for a remote zone node.
+func (zic *ZoneInterconnectHandler) deleteStaleStaticRoutes(node *corev1.Node) error {
+	var v4TransitSwitchIP, v6TransitSwitchIP, v4JoinIP, v6JoinIP *net.IPNet
+
+	nodeTransitSwitchPortIPs, err := util.ParseNodeTransitSwitchPortAddrs(node)
+	if err != nil || len(nodeTransitSwitchPortIPs) == 0 {
+		return fmt.Errorf("failed to get the node transit switch port Ips : %v", err)
+	}
+	for _, ip := range nodeTransitSwitchPortIPs {
+		if utilnet.IsIPv6CIDR(ip) {
+			v6TransitSwitchIP = ip
+		} else if utilnet.IsIPv4CIDR(ip) {
+			v4TransitSwitchIP = ip
+		} else {
+			return fmt.Errorf("invalid transit switch IP: %s", ip)
+		}
+	}
+
+	nodeGRPIPs, err := util.ParseNodeGatewayRouterJoinAddrs(node, zic.GetNetworkName())
+	if util.IsAnnotationNotSetError(err) {
+		var err1 error
+		// this is for backwards compatibility
+		nodeGRPIPs, err1 = util.ParseNodeGatewayRouterLRPAddrs(node)
+		if err1 != nil {
+			return fmt.Errorf("failed to parse node %s Gateway router LRP Addrs annotation %w", node.Name, err1)
+		}
+	}
+	for _, ip := range nodeGRPIPs {
+		if utilnet.IsIPv6CIDR(ip) {
+			v6JoinIP = ip
+		} else if utilnet.IsIPv4CIDR(ip) {
+			v4JoinIP = ip
+		} else {
+			return fmt.Errorf("invalid join switch IP: %s", ip)
+		}
+	}
+
+	nodeSubnets, err := util.ParseNodeHostSubnetAnnotation(node, zic.GetNetworkName())
+	if err != nil {
+		return fmt.Errorf("failed to parse node %s subnets annotation %w", node.Name, err)
+	}
+	p := func(lrsr *nbdb.LogicalRouterStaticRoute) bool {
+		if utilnet.IsIPv6String(lrsr.Nexthop) {
+			nodeSubnet, _ := util.MatchFirstIPNetFamily(true, nodeSubnets)
+			return lrsr.ExternalIDs["ic-node"] == node.Name && lrsr.ExternalIDs[types.NetworkExternalID] == zic.GetNetworkName() &&
+				(lrsr.Nexthop != v6TransitSwitchIP.IP.String() || (nodeSubnet != nil && lrsr.IPPrefix != nodeSubnet.String() &&
+					lrsr.IPPrefix != v6JoinIP.IP.String()+util.GetIPFullMaskString(v6JoinIP.IP.String())))
+		} else if !utilnet.IsIPv6String(lrsr.Nexthop) {
+			nodeSubnet, _ := util.MatchFirstIPNetFamily(false, nodeSubnets)
+			return lrsr.ExternalIDs["ic-node"] == node.Name && lrsr.ExternalIDs[types.NetworkExternalID] == zic.GetNetworkName() &&
+				(lrsr.Nexthop != v4TransitSwitchIP.IP.String() || (nodeSubnet != nil && lrsr.IPPrefix != nodeSubnet.String() &&
+					lrsr.IPPrefix != v4JoinIP.IP.String()+util.GetIPFullMaskString(v4JoinIP.IP.String())))
+		}
+		return false
+	}
+	err = libovsdbops.DeleteLogicalRouterStaticRoutesWithPredicate(zic.nbClient, zic.networkClusterRouterName, p)
+	if err != nil && !errors.Is(err, libovsdbclient.ErrNotFound) {
```

Comment:
I don't think that `DeleteLogicalRouterStaticRoutesWithPredicate` returns `libovsdbclient.ErrNotFound` if there was nothing to delete.

Reply:
yaa, you are right , ErrNotFound is set to false at operationModel.